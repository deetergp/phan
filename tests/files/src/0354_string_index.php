<?php

class A354 {
    /** @var string */
    public $foo = 'aaa';

    public function test() {
        $this->foo[1] = 'x';
        intdiv($this->foo, 2);
        $this->foo[] = 'x';  // should warn
        $x = 'aaa';
        $x[2] = '1';
        echo count($x);  // should warn.
        $x[] = 'c';  // should warn
        $y = 'aaa';
        $y['offset'] = 'c';
    }

    public function testFetch() {
        $str = 'abc';
        $x = $str[0];
        $y = $str[[]];  // wrong
        $z = $str['offset'];  // wrong
        $a = $str[null];  // wrong
        $arr = ['offset' => 'c'];
        $b = $arr['offset'];
        $c = $arr[[]];  // wrong
        $c = $arr[null];  // wrong
        $c = $arr[0];  // Phan isn't able to tell that this is wrong.
    }
}
