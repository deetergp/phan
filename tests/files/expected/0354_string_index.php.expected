%s:9 PhanTypeMismatchArgumentInternal Argument 1 (numerator) is string but \intdiv() takes int
%s:10 PhanTypeMismatchDimEmpty Assigning to an empty array index of a value of type string, but expected the index to exist and be of type int
%s:10 PhanTypeMismatchProperty Assigning string[] to property but \A354::foo is string
%s:13 PhanTypeMismatchArgumentInternal Argument 1 (var) is string but \count() takes \Countable|array
%s:14 PhanTypeMismatchDimEmpty Assigning to an empty array index of a value of type string, but expected the index to exist and be of type int
%s:16 PhanTypeMismatchDimAssignment When appending to a value of type string, found an array access index of type string, but expected the index to be of type int
%s:16 PhanTypeMismatchDimFetch When fetching an array index from a value of type string, found an array index of type string, but expected the index to be of type int
%s:22 PhanTypeMismatchDimFetch When fetching an array index from a value of type string, found an array index of type array, but expected the index to be of type int
%s:23 PhanTypeMismatchDimFetch When fetching an array index from a value of type string, found an array index of type string, but expected the index to be of type int
%s:24 PhanTypeMismatchDimFetch When fetching an array index from a value of type string, found an array index of type null, but expected the index to be of type int
%s:27 PhanTypeMismatchDimFetch When fetching an array index from a value of type string[], found an array index of type array, but expected the index to be of type int|string
%s:28 PhanTypeMismatchDimFetch When fetching an array index from a value of type string[], found an array index of type null, but expected the index to be of type int|string
